pkg/storegateway/bucket_index_reader.go

```diff
@@ -268,6 +228,95 @@ func (r *bucketIndexReader) expandedPostings(ctx context.Context, ms []*labels.M
 	return ps, nil
 }
 
+// toPostingGroups returns a set of labels for which to look up postings lists. It guarantees that
+// each postingGroup's addKeys and removeKeys exist in the index. The order of the returned labels
+// is the same as iterating through the posting groups and for each group adding all addKeys and then adding all removeKeys.
+func toPostingGroups(ms []*labels.Matcher, indexhdr indexheader.Reader) ([]postingGroup, []labels.Label, error) {
+	var (
+		postingGroups     = make([]postingGroup, 0, len(ms))
+		lazyPostingGroups []lazyPostingGroup
+		allRequested      = false
+		hasAdds           = false
+	)
+
+	// NOTE: Derived from tsdb.PostingsForMatchers.
+	for _, m := range ms {
+		// Each group is separate to tell later what postings are intersecting with what.
+		pg, lazyPg, isLazy := toPostingGroup(m)
+		if isLazy {
+			lazyPostingGroups = append(lazyPostingGroups, lazyPg)
+		} else {
+			postingGroups = append(postingGroups, pg)
+		}
+	}
+
+	// Next we check whether the posting groups won't select an empty set of postings.
+	// We start with the ones that have a known set of values because it's less expensive to check them in
+	// the index header.
+	numKeys := 0
+	for i, pg := range postingGroups {
+		var err error
+		pg, err = pg.filterNonExistentKeys(indexhdr)
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "filtering posting group")
+		}
+		postingGroups[i] = pg
+
+		// If this groups adds nothing, it's an empty group. We can shortcut this, since intersection with empty
+		// postings would return no postings anyway.
+		// E.g. label="non-existing-value" returns empty group.
+		if !pg.addAll && len(pg.addKeys) == 0 {
+			return nil, nil, nil
+		}
+
+		allRequested = allRequested || pg.addAll
+		hasAdds = hasAdds || len(pg.addKeys) > 0
+
+		numKeys += len(pg.addKeys)
+		numKeys += len(pg.removeKeys)
+	}
+
+	// We continue with the posting groups that require to scan all the label values from the index.
+	for _, lazyPg := range lazyPostingGroups {
+		pg, err := lazyPg.toPostingGroup(indexhdr)
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "toPostingGroup")
+		}
+		postingGroups = append(postingGroups, pg)
+		// If this groups adds nothing, it's an empty group. We can shortcut this, since intersection with empty
+		// postings would return no postings anyway.
+		// E.g. label="non-existing-value" returns empty group.
+		if !pg.addAll && len(pg.addKeys) == 0 {
+			return nil, nil, nil
+		}
+
+		allRequested = allRequested || pg.addAll
+		hasAdds = hasAdds || len(pg.addKeys) > 0
+
+		numKeys += len(pg.addKeys)
+		numKeys += len(pg.removeKeys)
+	}
```

Comment:
The code in these two loops is almost identical, the difference is that on the `postingGroups` we do the `filterNonExistentKeys()`, while on lazy ones we do `toPostingGroup()`, I would suggest switching back to just one type, having a `lazy bool` flag on them, and calling a single `postingsGroup.prepare()` method, which would do the filtering for `lazy=false` and label iterations for `lazy=true`, reducing the amount of logic in this method. 

If we put both types into same slice, we can sort them afterwards putting the lazy ones at the end, and sorting the non-lazy ones by the amount of `addKeys` they have before filtering, as those are most likely to shortcut (consider a year-long query for `pod="foo"` that was alive just a few moments in that year.)

Reply:
I went with a separate data structure because I didn't want to mix two different usage patterns/behaviours in the same data structure (`postingsGroup`).

I'm thinking that we can unify them behind an interface instead of conjoining their struct fields. 

> If we put both types into same slice, we can sort them afterwards putting the lazy ones at the end, and sorting the non-lazy ones by the amount of addKeys they have before filtering, as those are most likely to shortcut

we'll likely have just one key for most non-lazy ones. We get multiple only when it's a regex set matcher (`pod=~"compactor-(1|2|3|4|5)"`). So this will be an optimization for when the set matcher has a lot of variations _and_ there is a non-existent label in another matcher. You think we should still optimize for this?

Reply:
i can try with a single data structure. it might not be that bad. i will push and let you know

Reply:
I would avoid using interfaces here, since they'd only add indirection and extra allocation here. And we don't need _that much_ polymorphism, we just need _bimorphism_. That's why I suggested a bool.

> mix two different usage patterns/behaviours in the same data structure

Instead of that, we now have two duplicated code paths with just one method call difference between them (plus, we're allocating an extra slice for the lazy ones).


> You think we should still optimize for this?

I think that since we're touching this, why not making it better as it barely requires more code. (if not less, actually: the amount of code you save by not having to iterate through two similar slices is compensated by the sorting function).